numpy/core/include/numpy/arrayscalars.h

```diff
@@ -135,7 +135,13 @@ typedef struct {
 } PyScalarObject;
 
 #define PyStringScalarObject PyStringObject
-#define PyUnicodeScalarObject PyUnicodeObject
+#define PyStringScalarObject PyStringObject
+typedef struct {
+        /* note that the PyObject_HEAD macro lives right here */
+        PyUnicodeObject base;
+        Py_UCS4 *obval;
+} PyUnicodeScalarObject;
```

Comment:
Do we need to worry about ABI changes here?

Reply:
We could check sizeof(PyUnicodeObject) == sizeof(PyUnicodeScalarObject.base) at import

Also this should hit the release note.

Reply:
@mattip, your comment suggests an attempt at resilience to cpython ABI changes, which is something we don't care about because we already require recompilation against every version of python.

What hameer is point out is that anyone allocating subclassing our scalars in C (why would anyone do that), or allocating a PyUnicodeScalarObject on the stack (impossible to do without poking at CPython internals) will notice that our struct changed size.

Reply:
It was mentioned in the meeting that there's an API change here if you access the slots as well.

Reply:
The slots were all private CPython implementation details, so anyone accessing them was already fending for themselves

Reply:
Ah, so API is no issue, and I fully agree that the ABI (size change) is absurd to give much heed to... We could note it in the release notes, but thinking about it more, I am not even sure that makes any sense.

Reply:
If they’re private, I’m okay with putting this in.